Fixes for Mediatek mt818X mt8195 patforms

andrew-mtk opened the pull request:
Fixes for Mediatek mt818X mt8195 patforms

andrew-mtk committed:
soc/mediatek/adsp: Fix ELF .sof_entry section flags

Add 'ax' flags for .sof_entry ELF section so it merges properly with
.z_xtensa_vectors ELF section

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andrew-mtk committed:
soc/mediatek/adsp: Fix enable/disable timer interrups for MT8188

On MT8188 platform timer interrupts must also be enabled/disabled in
MTK_ADSP_IRQ_EN register in addition to xtensa_irq_enable(),
xtensa_irq_disable()

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andrew-mtk committed:
drivers/timer: Fix timer clock source initialization for MT818X

Fix timer clock source initialization for MT8188, MT8186 platform.
Add 26Mhz define that is specific for MT818X.

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andrew-mtk committed:
boards/mediatek: Include AFE dts from the board device tree

Include AFE dai definitions from the board device tree for MT8188
and MT8195 platforms

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andrew-mtk committed:
drivers/timer: Fix timer IRQ enable bit on MT818X and MT8195

MT818X and MT8195 use the same timer IRQ enable bit as MT8196

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andyross commented on soc/mediatek/mt8xxx/soc.c:
```diff
@@ -173,7 +173,7 @@ __asm__(".align 4\n\t"
  * needlessly.
  */
-__asm__(".pushsection .sof_entry.text\n\t"
+__asm__(".pushsection .sof_entry.text, \"ax\"\n\t"
```
Hm... not sure about this?  This section can't be part of any linker defaults, it really needs to be placed precisely at the beginning of the image to conform with the entry point set by the kernel driver.

It's not wrong to set the flags, but if it's not going where you need it the bug is in the linker script.

andyross commented on soc/mediatek/mt8xxx/irq.c:
```diff
@@ -46,6 +46,11 @@ void z_soc_irq_enable(unsigned int irq)
 		xtensa_irq_enable(irq);
+#if defined(CONFIG_SOC_MT8188)
+		if (irq < 25) {
+			*(volatile uint32_t*)(DT_REG_ADDR(DT_NODELABEL(intc2))) |= BIT(irq);
```
Can't this code go in `drivers/interrupt_controller/intc_mtk_adsp.c` instead?  This file is just for managing the routing of interrupts to different managers.  Zephyr embarrassingly lacks a good top-level framework for interrupt controllers, but it **does** have a reasonable API for the individual bits of hardware.

andrew-mtk replied:
> This file is just for managing the routing of interrupts to different managers.

This is why I put these changes over here. On MT818X platform Xtensa architectural vectors INT0 to INT24 are routed through external interrupt controller and must be enabled with both: `xtensa_irq_enable()` and corresponding bit in `mediatek,adsp_intc`

andyross replied:
Fair enough.  Sort of an angels on a pin head kind of argument as to whether that "external interrupt controller" deserves its own driver if it's 1:1 with the architectural interrupts.  Not the worst things in the world to leave it at the SOC level as long as it's this simple, though in general per-platform #if'ery is frowned on.

andyross reviewed (commented):
Handful of nitpicks/questions.  But just to repeat from the SOF side: it's so amazingly nice to (1) see vendor attention to this code and (2) that apparently I didn't mess it up too badly! :)

andrew-mtk commented:
@andyross, could you ping somebody who can review this?

nashif commented:
> not really, that platform target should build without SOF, all DSP target we have in tree build without SOF.

SOF is optional and zephyr has no dependency on SOF, it is the other way round.

andrew-mtk committed:
soc/mediatek/adsp: Add cpuclk driver for mt8188

Add cpuclk driver for mt8188 platform. Note that the cpuclk driver is
not yet ported, it works only with mt8188.

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andrew-mtk committed:
boards/mediatek: Add configs for mt8195, mt818X

Add configs for mt8195 and mt818X platforms to build them outside of SOF
environment and run by twister

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andrew-mtk committed:
drivers/timer: Fix build error when run by twister

Fix undefined reference to z_sys_timer_irq_for_test

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andrew-mtk committed:
west.yml: Update hal_xtensa revision

Update hal_xtensa revision to add HAL integration for mt8186/88
and mt8196 DSP processors.

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andrew-mtk commented:
@nashif , could you check it now?

andyross reviewed (approved):
This all looks great here.  None of the changes are particularly controversial, apologies that this hasn't merged yet.

I'm  a little worried about @nashif's inability to get these to build vs. the SDK, which should have toolchains integrated.  But it's possible that something has broken (obviously very little work against the DSPs happens with gcc).  I'll try these when I get a change, but surely nothing in these changes will have broken something at the toolchain level.  It's probably just broken in-tree already and shouldn't block merge.

andrew-mtk commented:
@andyross , thanks for your review. Do you by any chance have mt8188 hardware to try SOF firmware build?
